AWS: trust region if found from AWS metadata

justinsb opened the pull request:
```release-note
AWS: trust region if found from AWS metadata
```

Means we can run in newly announced regions without a code change.

We don't register the ECR provider in new regions, so we will still need
a code change for now.

Fix #35014

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/38880)
<!-- Reviewable:end -->

justinsb committed:
AWS: trust region if found from AWS metadata

Means we can run in newly announced regions without a code change.

We don't register the ECR provider in new regions, so we will still need
a code change for now.

This also means we do trust config / instance metadata, and don't reject
incorrectly configured zones.

Fix #35014

derekwaynecarr commented:
This is extremely useful.  Thanks for putting this together.

justinsb commented:
Thanks @derekwaynecarr - you can see the ECR issue as well ...  with this PR we won't have to register new regions in our code, but we still won't be able to use ECR in a new region unless we are running in that region.

To tackle that I think we would have to either change it so that we could build docker configs on demand, or just issue a DescribeRegions call and automatically register all of those regions.  Unsure which is better right now.

chrislovecnm commented:
>  you can see the ECR issue as well ... with this PR we won't have to register new regions in our code, but we still won't be able to use ECR in a new region unless we are running in that region.
> To tackle that I think we would have to either change it so that we could build docker configs on demand, or just issue a DescribeRegions call and automatically register all of those regions. Unsure which is better right now.

Does this remove functionality that we have now with ECR?

chrislovecnm commented on pkg/credentialprovider/aws/aws_credentials.go:
```diff
@@ -100,21 +80,20 @@ type ecrProvider struct {
 var _ credentialprovider.DockerConfigProvider = &ecrProvider{}
 
-// Init creates a lazy provider for each AWS region, in order to support
+// RegisterCredentialsProvider registers a credential provider for the specified region.
```
Do we have this documented anywhere?  If so do we need to update the docs?

chrislovecnm reviewed (commented):
Quick question

chrislovecnm commented:
@erictune a couple of questions, and since there actually is an active @kubernetes/sig-aws-misc group that is active, should we be doing code reviews now?

justinsb commented:
> Does this remove functionality that we have now with ECR?

No, it just adds functionality: we automatically recognize the ECR region we are running in, even if it isn't in our hard-coded list.

chrislovecnm reviewed (approved):
Looks great ... @erictune can we get this in?  This is super important :)

chrislovecnm commented:
@erictune can I get a PR review :)

colhom commented:
fixes https://github.com/coreos/kube-aws/issues/198

zmerlynn commented:
@k8s-bot aws test this

zmerlynn commented:
/lgtm

justinsb commented:
@zmerlynn OK if I reapply LGTM after the rebase I had to do?

zmerlynn commented:
/lgtm

@justinsb yup

justinsb commented:
@k8s-bot gci gce e2e test this

chrislovecnm commented:
/approve

chrislovecnm commented:
Let's see if that works

chrislovecnm commented:
@justinb you can probably self approve

chrislovecnm commented:
Can we get this cherry picked into 1.5.x??

chrislovecnm commented:
@zmerlynn can we get another looks good to me?

zmerlynn commented:
/lgtm

grodrigues3 commented:
@k8s-bot test this

Tests are more than 96 hours old. Re-running tests.

deads2k commented:
Wow, this approval thing is going to be really frustrating.  Also, why would it suggest asking yourself for approval??

I trust @zmerlynn  to know what this does.

/approve

zmerlynn commented:
@deads2k: Looks like the comment has to *start* with /approved

deads2k commented:
/approve

smarterclayton commented:
/approve
